demo/agilebank/templates/k8srequiredprobes_template.yaml

```diff
@@ -0,0 +1,53 @@
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata:
+  name: k8srequiredprobes
+spec:
+  crd:
+    spec:
+      names:
+        kind: K8sRequiredProbes
+      validation:
+        openAPIV3Schema:
+          properties:
+            probes:
+              type: array
+              items:
+                type: string
+            probeTypes:
+              type: array
+              items:
+                type: string
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package k8srequiredprobes
+
+        probe_type_set = probe_types {
+          probe_types := {type | type := input.parameters.probeTypes[_]}
+        }
+
+        violation[{"msg": msg}] {
+          container := input.review.object.spec.containers[_]
+          probe := input.parameters.probes[_]
+          probe_is_missing(container, probe)
+          msg := get_violation_message(container, input.review, probe)
+        }
+
+        probe_is_missing(ctr, probe) = true {
+          not ctr[probe]
+        }
+
+        probe_is_missing(ctr, probe) = true {
+          probe_field_empty(ctr, probe)
+        }
+
+        probe_field_empty(ctr, probe) = true {
+          probe_fields := {field | ctr[probe][field]}
+          diff_fields := probe_type_set - probe_fields
+          count(diff_fields) == count(probe_type_set)
```

Comment:
if `"probeTypes": ["tcpSocket", "httpGet"]` in the constraint and `probe_fields:["httpGet"]`, then this would resolve to `true`. but that is not what you want right?

Reply:
diff_fields would be `{ "tcpSocket" }`; and so `count(diff_fields)` would be 1.
With probe_type_set equal to `{ "tcpSocket", "httpGet" }`, `count(probe_type_set)` would be 2. `count(diff_fields) == count(probe_type_set)` give me false, meaning one of the two probes was specified and there's no violation.